eng/download-source-built-archive.sh

```diff
@@ -0,0 +1,108 @@
+#!/bin/bash
+
+# Common helper functions for downloading archives based on versions in XML files
+
+# Get the repository root directory
+REPO_ROOT="$(cd "$(dirname "${BASH_SOURCE[0]}")/.." && pwd)"
+
+# Define the path to Versions.props once
+PACKAGE_VERSIONS_PATH="$REPO_ROOT/eng/Versions.props"
+
+# Helper to extract a property value from an XML file
+function GetXmlPropertyValue {
+  local propName="$1"
+  local filePath="$2"
+  local value=""
+  local line pattern
+  line=$(grep -m 1 "<$propName>" "$filePath" || :)
+  pattern="<$propName>(.*)</$propName>"
+  if [[ $line =~ $pattern ]]; then
+    value="${BASH_REMATCH[1]}"
+  fi
+  echo "$value"
+}
+
+# Helper to download a file with retries
+function DownloadWithRetries {
+  local url="$1"
+  local targetDir="$2"
+  (
+    cd "$targetDir" &&
+    for i in {1..5}; do
+      if curl -fL --retry 5 -O "$url"; then
+        return 0
+      else
+        case $? in
+          18)
+            sleep 3
+            ;;
+          *)
+            return 1
+            ;;
+        esac
+      fi
+    done
+    return 1
+  )
+}
+
+# Generic function to download an archive based on property name
+function DownloadArchive {
+  local label="$1"
+  local propertyName="$2"
+  local isRequired="$3"
+  local artifactsRid="$4"
+  local outputDir="$5"
+  local destinationFilenamePrefix="${6:-}"
+
+  local notFoundMessage="No $label found to download..."
+
+  local archiveVersion
+  archiveVersion=$(GetXmlPropertyValue "$propertyName" "$PACKAGE_VERSIONS_PATH")
+  if [[ -z "$archiveVersion" ]]; then
+    if [ "$isRequired" == true ]; then
+      echo "  ERROR: $notFoundMessage"
+      return 1
+    else
+      echo "  $notFoundMessage"
+      return 0
+    fi
+  fi
+
+  local archiveUrl
+  local artifactsBaseFileName="Private.SourceBuilt.Artifacts"
+  local prebuiltsBaseFileName="Private.SourceBuilt.Prebuilts"
+  local defaultArtifactsRid='centos.10-x64'
+
+  if [[ "$propertyName" == "MicrosoftNETSdkVersion" ]]; then
+    archiveUrl="https://ci.dot.net/public/source-build/$artifactsBaseFileName.$archiveVersion.$artifactsRid.tar.gz"
+  elif [[ "$propertyName" == *Prebuilts* ]]; then
+    archiveUrl="https://builds.dotnet.microsoft.com/source-built-artifacts/assets/$prebuiltsBaseFileName.$archiveVersion.$defaultArtifactsRid.tar.gz"
```

Comment:
Will this need to account for internal builds?  Should/can that be accounted for now?

Reply:
It will, yes. I will log a follow-up issue. There's not really a way to test this right now.